applications/crossbar/src/modules/cb_tasks.erl

```diff
@@ -0,0 +1,433 @@
+-spec ctp(cb_context:context()) -> cb_context:context().
+ctp(Context, ?HTTP_GET) ->
+    ctp(Context);
+ctp(Context, _) ->
+    Context.
+ctp(Context) ->
+    CTP = [{'to_json', ?JSON_CONTENT_TYPES}
+          ,{'to_binary', ?CSV_CONTENT_TYPES}
+          ],
+    cb_context:add_content_types_provided(Context, CTP).
+
+%%--------------------------------------------------------------------
+%% @public
+%% @doc
+%% Check the request (request body, query string params, path tokens, etc)
+%% and load necessary information.
+%% /tasks mights load a list of task objects
+%% /tasks/123 might load the task object 123
+%% Generally, use crossbar_doc to manipulate the cb_context{} record
+%% @end
+%%--------------------------------------------------------------------
+-spec validate(cb_context:context()) -> cb_context:context().
+-spec validate(cb_context:context(), path_token()) -> cb_context:context().
+validate(Context) ->
+    validate_tasks(Context, cb_context:req_verb(Context)).
+validate(Context, PathToken) ->
+    validate_task(Context, PathToken, cb_context:req_verb(Context)).
+
+-spec validate_tasks(cb_context:context(), http_method()) -> cb_context:context().
+validate_tasks(Context, ?HTTP_GET) ->
+    summary(Context);
+validate_tasks(Context, ?HTTP_PUT) ->
+    QS = cb_context:query_string(Context),
+    case {kz_json:get_ne_binary_value(?QS_CATEGORY, QS)
+         ,kz_json:get_ne_binary_value(?QS_ACTION, QS)
+         }
+    of
+        {?NE_BINARY, ?NE_BINARY} ->
+            validate_new_attachment(Context, is_content_type_csv(Context));
+        {_, _} -> cb_context:add_system_error('invalid request', Context)
+    end.
+
+-spec validate_task(cb_context:context(), path_token(), http_method()) -> cb_context:context().
+validate_task(Context, ?HELP, ?HTTP_GET) ->
+    lager:debug("starting discovery of task APIs"),
+    JObj = kz_json:from_list([{<<"tasks">>, kz_tasks:help()}]),
+    cb_context:setters(Context, [{fun cb_context:set_resp_status/2, 'success'}
+                                ,{fun cb_context:set_resp_data/2, JObj}
+                                ]);
+validate_task(Context, TaskId, ?HTTP_GET) ->
+    Context1 = read(TaskId, Context),
+    case cb_context:resp_status(Context1) == 'success'
+        andalso is_accept_csv(Context)
+    of
+        'false' -> Context1;
+        'true' ->
+            lager:debug("trying to fetch attachment for task ~s", [TaskId]),
+            AName = kz_tasks:attachment_name(TaskId),
+            load_csv_attachment(Context, TaskId, AName)
+    end;
+validate_task(Context, TaskId, ?HTTP_PATCH) ->
+    read(TaskId, Context);
+validate_task(Context, TaskId, ?HTTP_DELETE) ->
+    read(TaskId, Context).
+
+-spec validate_new_attachment(cb_context:context(), boolean()) -> cb_context:context().
+validate_new_attachment(Context, 'true') ->
+    [{_Filename, FileJObj}] = cb_context:req_files(Context),
+    CSVBinary = kz_json:get_value(<<"contents">>, FileJObj),
+    case kz_tasks:is_csv(CSVBinary) of
```

Comment:
Returning the count of rows from a function named `is_csv` is surprising. `is_*` are typically returning `boolean()`. Could you do `case is_csv(CSVBinary) andalso csv_rows(CSVBinary) of` and handle 'false', 0, and TotalRows clauses?